cgsmiles/sample.py

```diff
@@ -7,133 +7,240 @@
-        source_node = np.random.choice(self.current_open_bonds[bonding])
-        compl_bonding = _find_complementary_bonding_descriptor(bonding)
+        # 3. get a corresponding node; it may be that one descriptor is found on
+        #    several nodes
+        random.seed(a=seed)
+        source_node = random.choice(self.current_open_bonds[bonding])
+        # 4. get the complementary matching bonding descriptor
+        compl_bonding = find_complementary_bonding_descriptor(bonding)
+        # 5. pick a new fragment that has such bonding descriptor
+        random.seed(a=seed)
         fragname, target_node = random.choice(self.fragments_by_bonding[compl_bonding])
-        return (source_node, target_node, fragname, bonding, compl_bonding)
+        # 6. add the new fragment and do some book-keeping
+        correspondence = merge_graphs(molecule, self.fragment_dict[fragname])
+        molecule.add_edge(source_node,
+                          correspondence[target_node],
+                          bonding=(bonding, compl_bonding))
+        molecule.nodes[source_node]['bonding'].remove(bonding)
+        molecule.nodes[correspondence[target_node]]['bonding'].remove(compl_bonding)
+        self.current_open_bonds = find_open_bonds(molecule)
+        return molecule, fragname
+
+    def terminate_branch(self, molecule, fragname, fragid, seed=None):
+        """
+        Probabilistically terminate a branch by removing all
+        bonding descriptors from the last fragment.
 
-    def sample(self):
+        Parameters
+        ----------
+        molecule: nx.Graph
+            the molecule graph
+        fragname: str
+            the fragment by name to consider
+        fragid: int
+            the id of the fragment
+
+        Returns
+        -------
+        nx.Graph
+        """
+        term_prob = self.termination_probabilities.get(fragname, 0)
+        random.seed(a=seed)
+        # probability check for termination
+        if random.random() < term_prob:
+            # check if there are more open bonding descriptors
+            # if the number is the same as would get removed
+            # then we are not on a branch
+            active_bonds = nx.get_node_attributes(molecule, 'bonding')
+            target_nodes = [ node for node in active_bonds if molecule.nodes[node]['fragid'] == fragid]
+            if len(target_nodes) < len(active_bonds):
+                for node in target_nodes:
+                    del molecule.nodes[node]['bonding']
+                self.current_open_bonds = find_open_bonds(molecule)
+        return molecule
+
+    def sample(self, target_weight, seed=None):
         """
-        Sample one molecule given a fragment string and a target molecular
-        weigth.
+        From a list of cgsmiles fragment graphs generate a new random molecule
+        according by stitching them together.
+
+        Parameters
+        ----------
+        target_weight
+            the weight of the polymer to generate
```

Comment:
Maximum or minimum?
Example, target_weight is 10, current weight is 9. Do I add a weight 2 fragment or not?

Reply:
minimum; docstring is updated